aiounifi/models/traffic_rule.py

```diff
@@ -0,0 +1,167 @@
+"""Traffic rules as part of a UniFi network."""
+
+from dataclasses import dataclass
+from typing import Any, NotRequired, Self, TypedDict
+
+import orjson
+
+from .api import ApiItem, ApiRequestV2, TypedApiResponse
+
+
+class BandwidthLimit(TypedDict):
+    """Bandwidth limit type definition."""
+
+    download_limit_kbps: int
+    enabled: bool
+    upload_limit_kbps: int
+
+
+class PortRange(TypedDict):
+    """Port range type definition."""
+
+    port_start: int
+    port_stop: int
+
+
+class IPAddress(TypedDict):
+    """IP Address for which traffic rule is applicable type definition."""
+
+    ip_or_subnet: str
+    ip_version: str
+    port_ranges: list[PortRange]
+    ports: list[int]
+
+
+class IPRange(TypedDict):
+    """IP Range type definition."""
+
+    ip_start: str
+    ip_stop: str
+    ip_version: str
+
+
+class Schedule(TypedDict):
+    """Schedule to enable/disable traffic rule type definition."""
+
+    date_end: str
+    date_start: str
+    mode: str
+    repeat_on_days: list[str]
+    time_all_day: bool
+    time_range_end: str
+    time_range_start: str
+
+
+class TargetDevice(TypedDict):
+    """Target device to which the traffic rule applies."""
+
+    client_mac: NotRequired[str]
+    network_id: NotRequired[str]
+    type: str
+
+
+class TypedTrafficRule(TypedDict):
+    """Traffic rule type definition."""
+
+    _id: str
+    action: str
+    app_category_ids: list[str]
+    app_ids: list[str]
+    bandwidth_limit: BandwidthLimit
+    description: str
+    domains: list[str]
+    enabled: bool
+    ip_addresses: list[IPAddress]
+    ip_ranges: list[IPRange]
+    matching_target: str
+    network_ids: list[str]
+    regions: list[str]
+    schedule: Schedule
+    target_devices: list[TargetDevice]
+
+
+@dataclass
+class TrafficRuleRequest(ApiRequestV2):
+    """Data class with required properties of a traffic rule API request."""
+
+    def prepare_data(self, raw: bytes) -> TypedApiResponse:
+        """Put data, received from the unifi controller, into a TypedApiResponse."""
+        json_data = orjson.loads(raw)
+        return_data: TypedApiResponse = {}
+        return_data["meta"] = self.handle_error(json_data)
+        return_data["data"] = json_data
+        return return_data
+
+
+@dataclass
+class TrafficRuleToggleRequest(ApiRequestV2):
+    """Data class with required properties of a traffic rule toggle API request."""
+
+    def prepare_data(self, raw: bytes) -> TypedApiResponse:
+        """Put data, received from the unifi controller, into a TypedApiResponse."""
+        json_data = orjson.loads(raw)
+        return_data: TypedApiResponse = {}
+        return_data["meta"] = self.handle_error(json_data)
+        return_data["data"] = [json_data]
+        return return_data
```

Comment:
```suggestion
    def prepare_data(self, raw: bytes) -> TypedApiResponse:
        """Put data, received from the unifi controller, into a TypedApiResponse."""
        json_data = orjson.loads(raw)
        return_data: TypedApiResponse = {}
        return_data["meta"] = self.handle_error(json_data)
        return_data["data"] = json_data
        return return_data


@dataclass
class TrafficRuleToggleRequest(ApiRequestV2):
    """Data class with required properties of a traffic rule toggle API request."""

    def prepare_data(self, raw: bytes) -> TypedApiResponse:
        """Put data, received from the unifi controller, into a TypedApiResponse."""
        json_data = orjson.loads(raw)
        return_data: TypedApiResponse = {}
        return_data["meta"] = self.handle_error(json_data)
        return_data["data"] = [json_data]
        return return_data
```
I think this very well also can be part of the V2 base class, you only need to consider json_data being a list or not right? Does it being a list only depend on if it is a get or post?

Reply:
Yes, I only need to consider json_data being a list or not.
A GET results in a list.
A PUT results in a dict.

I moved this 'prepare_data' (now 'decode') to the V2 base class; this also enabled me to remove the other TrafficRule...Request classes. I will commit this in a few moments.